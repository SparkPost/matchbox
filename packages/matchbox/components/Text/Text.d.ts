export default Text;
declare const Text: any;
