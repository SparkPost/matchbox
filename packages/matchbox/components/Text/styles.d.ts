export function truncate(props: any): {
    overflow: string;
    textOverflow: string;
    whiteSpace: string;
};
export function lookslike(props: any): string;
