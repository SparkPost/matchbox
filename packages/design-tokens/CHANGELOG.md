# @sparkpost/design-tokens

## 7.1.1

### Patch Changes

- Fix missing dist
