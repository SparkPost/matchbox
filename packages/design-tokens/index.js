const tokens = require('./dist/index.common.js');
const meta = require('./dist/index.meta.js');

module.exports = {
  tokens,
  meta,
};
