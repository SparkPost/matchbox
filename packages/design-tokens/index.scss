@import './dist/scss/tokens.scss';
