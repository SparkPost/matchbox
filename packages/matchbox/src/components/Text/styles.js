import tokens from '@sparkpost/design-tokens/meta';
import _ from 'lodash';

export const truncate = (props) => {
  if (props.truncate) {
    return {
      overflow: 'hidden',
      textOverflow: 'ellipsis',
      whiteSpace: 'nowrap'
    };
  }
};

/**
 * Crops the text elements line height.
 * The element's vertical bounds will begin & end at the rendered font instead of line height.
 * This is an experimental prop
 * Calibre naturally has more spacing above than below, cropping is not perfect.
 */
export const crop = (props) => {
  const capHeight = 0.69; // Cap height optimized for Calibre
  const baseline = 2; // Calculated height will always be divisible by this baseline

  // Requires lineHeight and fontSize to reference a token
  if (
    !props.crop || !props.lineHeight || !props.fontSize ||
    !_.find(tokens, ({ name }) => name === `line-height-${props.lineHeight}`) ||
    !_.find(tokens, ({ name }) => name === `font-size-${props.fontSize}`)
  ) {
    return;
  }

  const lineHeight = _.find(tokens, ({ name }) => name === `line-height-${props.lineHeight}`).pixel_value.replace('px', '');
  const fontSize = _.find(tokens, ({ name }) => name === `font-size-${props.fontSize}`).pixel_value.replace('px', '');

  const desiredHeight = capHeight * Number(fontSize);
  const spaceToRemove = (Number(lineHeight) - desiredHeight);
  const spaceRoundedToBaseline = spaceToRemove - (spaceToRemove % baseline);

  return `
    transform: translateY(${spaceRoundedToBaseline / 2}px);

    &::before {
      content: '';
      display: block;
      height: 0;
      width: 0;
      margin-top: -${spaceRoundedToBaseline}px;
    }
  `;
};
