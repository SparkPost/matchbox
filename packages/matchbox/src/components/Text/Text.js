import styled from 'styled-components';
import { color, space, typography, compose } from 'styled-system';
import { truncate, looksLike } from './styles';
import PropTypes from 'prop-types';

const system = compose(color, space, typography);

const Text = styled('p')`
  ${system}
  ${truncate}
  ${looksLike}
`;

Text.propTypes = {
  as: PropTypes.elementType.isRequired,
  children: PropTypes.node.isRequired,
  looksLike: PropTypes.oneOf(['h1', 'h2', 'h3', 'h4', 'h5', 'h6', 'p']),
};

Text.defaultProps = {
  as: 'p',
};

Text.displayName = 'Text';
export default Text;
