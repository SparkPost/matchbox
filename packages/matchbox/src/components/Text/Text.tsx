import React from 'react';
import styled from 'styled-components';
import {
  color,
  ColorProps,
  space,
  SpaceProps,
  typography,
  TypographyProps,
  compose,
} from 'styled-system';
import type * as Polymorphic from '../../helpers/types';
import { truncate, lookslike } from './styles';

const system = compose(color, space, typography);

const StyledText = styled.p`
  ${system}
  ${truncate}
  ${lookslike}
`;

export type TextProps = {
  'data-id'?: string;
  children: React.ReactNode;
  truncate?: boolean;
  looksLike?: 'h1' | 'h2' | 'h3' | 'h4' | 'h5' | 'h6' | 'p';
};

type PolymorphicText = Polymorphic.ForwardRefComponent<
  'p',
  TextProps & ColorProps & SpaceProps & TypographyProps
>;

const Text = React.forwardRef(function Text(props, ref) {
  const { as = 'p', looksLike, children, truncate, ...rest } = props;

  return (
    <StyledText as={as} $looksLike={looksLike} $truncate={truncate} ref={ref} {...rest}>
      {children}
    </StyledText>
  );
}) as PolymorphicText;

Text.displayName = 'Text';
export default Text;
