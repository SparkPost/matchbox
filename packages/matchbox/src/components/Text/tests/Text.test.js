import React from 'react';
import Text from '../Text';
import { tokens } from '@sparkpost/design-tokens';
import 'jest-styled-components';

describe('Text', () => {
  it('it should render correctly', () => {
    const wrapper = global.renderStyled(<Text id="text-id">Text</Text>);
    expect(wrapper).toMatchSnapshot();
  });

  it('it should truncate', () => {
    const wrapper = global.renderStyled(<Text truncate>Text</Text>);
    expect(wrapper).toMatchSnapshot();
  });

  describe('looksLike renders correctly', () => {
    it('it should render visually as the looksLike prop while html should match the as prop ', () => {
      const wrapper = global.mountStyled(
        <Text as="h3" looksLike="h1">
          Text
        </Text>,
      );
      expect(wrapper).toHaveStyleRule('font-size', tokens.fontSize_700);
    });
  });
});
