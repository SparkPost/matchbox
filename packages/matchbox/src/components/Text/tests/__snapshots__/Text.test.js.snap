// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Text cropping it should crop when provided a size and line height 1`] = `
.c0 {
  font-size: 1rem;
  line-height: 1.5rem;
  -webkit-transform: translateY(5px);
  -ms-transform: translateY(5px);
  transform: translateY(5px);
}

.c0::before {
  content: '';
  display: block;
  height: 0;
  width: 0;
  margin-top: -10px;
}

<p
  class="c0"
  font-size="400"
>
  Text
</p>
`;

exports[`Text cropping it should not crop with an unsupported line height token 1`] = `
.c0 {
  font-size: 1rem;
  line-height: 10px;
}

<p
  class="c0"
  font-size="400"
>
  Text
</p>
`;

exports[`Text cropping it should not crop with an unsupported size token 1`] = `
.c0 {
  font-size: 10px;
  line-height: 1.5rem;
}

<p
  class="c0"
  font-size="10px"
>
  Text
</p>
`;

exports[`Text it should render correctly 1`] = `
<p
  class=""
  id="text-id"
>
  Text
</p>
`;

exports[`Text it should truncate 1`] = `
.c0 {
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
}

<p
  class="c0"
>
  Text
</p>
`;
